go-controller/pkg/observability/observability.go

```diff
@@ -0,0 +1,210 @@
+const (
+	DropSamplingID = iota + 1
+	ACLNewTrafficSamplingID
+	ACLEstTrafficSamplingID
+)
+
+// temporary const, until we have dynamic config
+const DefaultObservabilityCollectorSetID = 28
+
+// this is inferred from nbdb schema, check Sample_Collector.id
+const maxCollectorID = 255
+
+// collectorConfig holds the configuration for a collector.
+// It is allowed to set different probabilities for every feature.
+// collectorSetID is used to set up sampling via OVSDB.
+type collectorConfig struct {
+	collectorSetID int
+	// probability in percent, 0 to 100
+	featuresProbability map[libovsdbops.SampleFeature]int
+}
+
+type Manager struct {
+	nbClient   libovsdbclient.Client
+	sampConfig *libovsdbops.SamplingConfig
+	// nbdb Collectors have probability. To allow different probabilities for different features,
+	// multiple nbdb Collectors will be created, one per probability.
+	// getCollectorKey() => collector.UUID
+	dbCollectors map[string]string
+	// Only maxCollectorID collectors are allowed, each should have unique ID.
+	// this set is tracking already assigned IDs.
+	takenCollectorIDs sets.Set[int]
+}
+
+func NewManager(nbClient libovsdbclient.Client) *Manager {
+	return &Manager{
+		nbClient:          nbClient,
+		dbCollectors:      make(map[string]string),
+		takenCollectorIDs: sets.New[int](),
+	}
+}
+
+func (m *Manager) SamplingConfig() *libovsdbops.SamplingConfig {
+	return m.sampConfig
+}
+
+func (m *Manager) Init() error {
+	if err := m.setSamplingAppIDs(); err != nil {
+		return err
+	}
+	if err := m.setDbCollectors(); err != nil {
+		return err
+	}
+	featuresConfig, err := m.addCollector(&collectorConfig{
+		collectorSetID: DefaultObservabilityCollectorSetID,
+		featuresProbability: map[libovsdbops.SampleFeature]int{
+			libovsdbops.EgressFirewall:     100,
+			libovsdbops.NetworkPolicy:      100,
+			libovsdbops.AdminNetworkPolicy: 100,
+			libovsdbops.Multicast:          100,
+			libovsdbops.UDNIsolation:       100,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	m.sampConfig = libovsdbops.NewSamplingConfig(featuresConfig)
+	return nil
+}
+
+func (m *Manager) setDbCollectors() error {
+	clear(m.dbCollectors)
+	collectors, err := libovsdbops.FindSampleCollectorWithPredicate(m.nbClient, func(collector *nbdb.SampleCollector) bool {
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("error getting sample collectors: %w", err)
+	}
+	for _, collector := range collectors {
+		m.dbCollectors[getCollectorKey(collector.SetID, collector.Probability)] = collector.UUID
+		m.takenCollectorIDs.Insert(collector.SetID)
+	}
+	return nil
+}
+
+// Cleanup must be called when observability is no longer needed.
+// It will return an error if some samples still exist in the db.
+// This is expected, and Cleanup may be retried on the next restart.
+func Cleanup(nbClient libovsdbclient.Client) error {
+	// Do the opposite of init
+	err := libovsdbops.DeleteSamplingAppsWithPredicate(nbClient, func(app *nbdb.SamplingApp) bool {
+		return true
+	})
```

Comment:
I had a comment above about "return true" predicates.  But I see now you use it in more places.  Is it an agreed way of doing things?

Reply:
I also found this rather strange. It seems a limitation of the `modelClient`.

Reply:
for the previous comment to get all objects, we use `nbCleint.List`, and I updated it.
But to delete, I don't think we have functions like deleteAll, as libovsdb doesn't provide plain Delete function, only conditional delete https://github.com/ovn-org/libovsdb/blob/main/client/api.go#L16

Reply:
OK, I think I would've added a libovsdb.DeleteAllSamplingApps() wrapper but if the current code is acceptable that's fine by me too.